File: CSharp/SpaceVIL/SpaceVIL/Items/TabView.cs
using System;
using System.Collections.Generic;
using SpaceVIL.Common;
using SpaceVIL.Core;
using SpaceVIL.Decorations;

namespace SpaceVIL
{
    /// <summary>
    /// TabView is a special container designed to group items within specific tab page.
    /// <para/> TabView cannot receive any events, 
    /// so TabView is always in the SpaceVIL.Core.ItemStateType.Base state.
    /// </summary>
    public class TabView : VerticalStack
    {
        static int count = 0;

        /// <summary>
        /// Setting the width policy of tabs. 
        /// Can be Fixed (tab shape not changes its size) or 
        /// Expand (tab shape is stretched to all available space).
        /// </summary>
        /// <param name="policy">Width policy of tab shapes.</param>
        public void SetTabPolicy(SizePolicy policy)
        {
            _tabBar.SetContentPolicy(policy);
        }

        /// <summary>
        /// Getting the current width policy of tabs. 
        /// Can be Fixed (tab shape not changes its size) or 
        /// Expand (tab shape is stretched to all available space).
        /// </summary>
        /// <returns>Width policy of tab shapes.</returns>
        public SizePolicy GetTabPolicy()
        {
            return _tabBar.GetContentPolicy();
        }

        private TabBar _tabBar;
        private Frame _viewArea;

        /// <summary>
        /// Getting all existing tabs in TabView.
        /// </summary>
        /// <returns>Tabs as List&lt;SpaceVIL.Tab&gt;.</returns>
        public List<Tab> GetTabs()
        {
            return _tabBar.GetTabs();
        }

        /// <summary>
        /// Default TabView constructor.
        /// </summary>
        public TabView()
        {
            SetItemName("TabView_" + count++);
            _tabBar = new TabBar();
            _viewArea = new Frame();
            SetStyle(DefaultsService.GetDefaultStyle(typeof(SpaceVIL.TabView)));
        }

        protected internal override bool GetHoverVerification(float xpos, float ypos)
        {
            return false;
        }

        /// <summary>
        /// Initializing all elements in the TabView.
        /// <para/> Notice: This method is mainly for overriding only. SpaceVIL calls 
        /// this method if necessary and no need to call it manually.
        /// </summary>
        public override void InitElements()
        {
            AddItems(_tabBar, _viewArea);
        }

        /// <summary>
        /// Selecting a tab by its Tab instance.
        /// </summary>
        /// <param name="tab">Tab to select.</param>
        public void SelectTab(Tab tab)
        {
            _tabBar.SelectTab(tab);
        }

        /// <summary>
        /// Selecting a tab by its index. 
        /// <para/> If index is out of range then this method does nothing.
        /// </summary>
        /// <param name="index">Index of a tab.</param>
        public void SelectTab(int index)
        {
            _tabBar.SelectTab(index);
        }

        /// <summary>
        /// Selecting a tab by its name. 
        /// <para/> Tab name is ID of UI element. Tab name is uniq but tab text is not.
        /// <para/> Explanation: the tab text can be a web address, 
        /// and there can be several open tabs with the same address.
        /// </summary>
        /// <param name="tabName">Tab name.</param>
        public void SelectTabByName(String tabName)
        {
            _tabBar.SelectTabByName(tabName);
        }

        /// <summary>
        /// Selecting a first-found tab by its text. 
        /// <para/> Tab name is ID of UI element. Tab name is uniq but tab text is not.
        /// <para/> Explanation: the tab text can be a web address, 
        /// and there can be several open tabs with the same address.
        /// </summary>
        /// <param name="tabText">Tab text.</param>
        public void SelectTabByText(String tabText)
        {
            _tabBar.SelectTabByText(tabText);
        }

        /// <summary>
        /// Getting the index of the current selected tab.
        /// </summary>
        /// <returns>Index of the current selected tab.</returns>
        public int GetSelectedTabIndex()
        {
            return _tabBar.GetSelectedTabIndex();
        }

        /// <summary>
        /// Getting the current selected tab.
        /// </summary>
        /// <returns>Current selected tab as SpaceVIL.Tab.</returns>
        public Tab GetSelectedTab()
        {
            return _tabBar.GetSelectedTab();
        }

        /// <summary>
        /// Getting a page of a tab. 
        /// </summary>
        /// <param name="tab">Tab as SpaceVIL.Tab.</param>
        /// <returns>Page of a tab as SpaceVIL.Frame.</returns>
        public Frame GetTabFrame(Tab tab)
        {
            return _tabBar.GetTabFrame(tab);
        }

        /// <summary>
        /// Getting content of a page by its tab.
        /// </summary>
        /// <param name="tab">Tab as SpaceVIL.Tab.</param>
        /// <returns>Page content as List&lt;SpaceVIL.IBaseItem&gt;.</returns>
        public List<IBaseItem> GetTabContent(Tab tab)
        {
            return _tabBar.GetTabContent(tab);
        }

        /// <summary>
        /// Adding a new tab to the TabView.
        /// </summary>
        /// <param name="tab">The new tab as SpaceVIL.Tab.</param>
        public void AddTab(Tab tab)
        {
            if (_tabBar.TabMapView.ContainsKey(tab))
            {
                return;
            }

            _tabBar.TabMapView.Add(tab, tab.View);
            _viewArea.AddItem(tab.View);
            _tabBar.AddItem(tab);
            tab.EventMousePress += (sender, args) =>
            {
                UpdateLayout();
            };
            tab.EventTabRemove += () =>
            {
                RemoveTab(tab);
                UpdateLayout();
            };
        }

        /// <summary>
        ///  Adding new tabs to the TabView.
        /// </summary>
        /// <param name="tabs">Tab sequence as SpaceVIL.Tab.</param>
        public void AddTabs(params Tab[] tabs)
        {
            foreach (Tab tab in tabs)
            {
                AddTab(tab);
            }
        }

        /// <summary>
        /// Removing the specified item from TabView.
        /// </summary>
        /// <param name="item"></param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public override bool RemoveItem(IBaseItem item)
        {
            Tab tmpTab = item as Tab;
            if (tmpTab != null)
            {
                if (_tabBar.TabMapView.ContainsKey(tmpTab))
                {
                    return RemoveTab(tmpTab);
                }
                return false;
            }
            else if (item.Equals(_tabBar))
            {
                _tabBar.RemoveAllTabs();
            }
            return base.RemoveItem(item);
        }

        /// <summary>
        /// Removing a tab by its name.
        /// <para/> Tab name is ID of UI element. Tab name is uniq but tab text is not.
        /// <para/> Explanation: the tab text can be a web address, 
        /// and there can be several open tabs with the same address.
        /// </summary>
        /// <param name="tabName">Tab name.</param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public bool RemoveTabByName(String tabName)
        {
            return _tabBar.RemoveTabByName(tabName);
        }

        /// <summary>
        /// Removing a first-found tab by its text.
        /// <para/> Tab name is ID of UI element. Tab name is uniq but tab text is not.
        /// <para/> Explanation: the tab text can be a web address, 
        /// and there can be several open tabs with the same address.
        /// </summary>
        /// <param name="tabText">Tab text.</param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public bool RemoveTabByText(String tabText)
        {
            return _tabBar.RemoveTabByText(tabText);
        }

        /// <summary>
        /// Removing a tab.
        /// </summary>
        /// <param name="tab">Tab as SpaceVIL.Tab.</param>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public bool RemoveTab(Tab tab)
        {
            return _tabBar.RemoveTab(tab);
        }

        /// <summary>
        /// Removing all existing tabs.
        /// </summary>
        /// <returns>True: if the removal was successful. 
        /// False: if the removal was unsuccessful.</returns>
        public bool RemoveAllTabs()
        {
            return _tabBar.RemoveAllTabs();
        }

        /// <summary>
        /// Adding an item to page by its tab name.
        /// <para/> Tab name is ID of UI element. Tab name is uniq but tab text is not.
        /// <para/> Explanation: the tab text can be a web address, 
        /// and there can be several open tabs with the same address.
        /// </summary>
        /// <param name="tabName">Tab name.</param>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        public void AddItemToTabByName(String tabName, IBaseItem item)
        {
            _tabBar.AddItemToTabByName(tabName, item);
        }

        /// <summary>
        /// Adding an item to page by its tab text.
        /// <para/> Tab name is ID of UI element. Tab name is uniq but tab text is not.
        /// <para/> Explanation: the tab text can be a web address, 
        /// and there can be several open tabs with the same address.
        /// </summary>
        /// <param name="tabText">Tab text.</param>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        public void AddItemToTabByText(String tabText, IBaseItem item)
        {
            _tabBar.AddItemToTabByText(tabText, item);
        }

        /// <summary>
        /// Adding an item to page by its tab.
        /// </summary>
        /// <param name="tab">Tab as SpaceVIL.Core.Tab.</param>
        /// <param name="item">Item as SpaceVIL.Core.IBaseItem.</param>
        public void AddItemToTab(Tab tab, IBaseItem item)
        {
            _tabBar.AddItemToTab(tab, item);
        }

        /// <summary>
        /// Setting style of the TabView.
        /// <para/> Inner styles: "tabbar", "viewarea".
        /// </summary>
        /// <param name="style">Style as SpaceVIL.Decorations.Style.</param>
        public override void SetStyle(Style style)
        {
            if (style == null)
            {
                return;
            }
            base.SetStyle(style);

            Style innerStyle = style.GetInnerStyle("tabbar");
            if (innerStyle != null)
            {
                _tabBar.SetStyle(innerStyle);
            }
            innerStyle = style.GetInnerStyle("viewarea");
            if (innerStyle != null)
            {
                _viewArea.SetStyle(innerStyle);
            }
        }
    }
}

File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/TabView.java
package com.spvessel.spacevil;

import com.spvessel.spacevil.Common.DefaultsService;
import com.spvessel.spacevil.Core.InterfaceBaseItem;
import com.spvessel.spacevil.Decorations.Style;
import com.spvessel.spacevil.Flags.SizePolicy;

import java.util.List;

/**
 * TabView is a special container designed to group items within specific tab
 * page.
 * <p>
 * TabView cannot receive any events, so TabView is always in the
 * com.spvessel.spacevil.Flags.ItemStateType.BASE state.
 */
public class TabView extends VerticalStack {
    private static int count = 0;

    /**
     * Setting the width policy of tabs. Can be FIXED (tab shape not changes its
     * size) or EXPAND (tab shape is stretched to all available space).
     * 
     * @param policy Width policy of tab shapes.
     */
    public void setTabPolicy(SizePolicy policy) {
        _tabBar.setContentPolicy(policy);
    }

    /**
     * Getting the current width policy of tabs. Can be FIXED (tab shape not changes
     * its size) or EXPAND (tab shape is stretched to all available space).
     * 
     * @return Width policy of tab shapes.
     */
    public SizePolicy getTabPolicy() {
        return _tabBar.getContentPolicy();
    }

    private TabBar _tabBar;
    private Frame _viewArea;

    /**
     * Getting all existing tabs in TabView.
     * 
     * @return Tabs as List&lt;com.spvessel.spacevil.Tab&gt;.
     */
    public List<Tab> getTabs() {
        return _tabBar.getTabs();
    }

    /**
     * Default TabView constructor.
     */
    public TabView() {
        setItemName("TabView_" + count++);
        _tabBar = new TabBar();
        _viewArea = new Frame();
        setStyle(DefaultsService.getDefaultStyle(TabView.class));
    }

    @Override
    protected boolean getHoverVerification(float xpos, float ypos) {
        return false;
    }

    /**
     * Initializing all elements in the TabView.
     * <p>
     * Notice: This method is mainly for overriding only. SpaceVIL calls this method
     * if necessary and no need to call it manually.
     */
    @Override
    public void initElements() {
        addItems(_tabBar, _viewArea);
    }

    /**
     * Selecting a tab by its Tab instance.
     * 
     * @param tab Tab to select.
     */
    public void selectTab(Tab tab) {
        _tabBar.selectTab(tab);
    }

    /**
     * Selecting a tab by its index.
     * <p>
     * If index is out of range then this method does nothing.
     * 
     * @param index Index of a tab.
     */
    public void selectTab(int index) {
        _tabBar.selectTab(index);
    }

    /**
     * Selecting a tab by its name.
     * <p>
     * Tab name is ID of UI element. Tab name is uniq but tab text is not.
     * <p>
     * Explanation: the tab text can be a web address, and there can be several open
     * tabs with the same address.
     * 
     * @param tabName Tab name.
     */
    public void selectTabByName(String tabName) {
        _tabBar.selectTabByName(tabName);
    }

    /**
     * Selecting a first-found tab by its text.
     * <p>
     * Tab name is ID of UI element. Tab name is uniq but tab text is not.
     * <p>
     * Explanation: the tab text can be a web address, and there can be several open
     * tabs with the same address.
     * 
     * @param tabText Tab text.
     */
    public void selectTabByText(String tabText) {
        _tabBar.selectTabByText(tabText);
    }

    /**
     * Getting the index of the current selected tab.
     * 
     * @return Index of the current selected tab.
     */
    public int getSelectedTabIndex() {
        return _tabBar.getSelectedTabIndex();
    }

    /**
     * Getting the current selected tab.
     * 
     * @return Current selected tab as com.spvessel.spacevil.Tab.
     */
    public Tab getSelectedTab() {
        return _tabBar.getSelectedTab();
    }

    /**
     * Getting a page of a tab.
     * 
     * @param tab Tab as com.spvessel.spacevil.Tab.
     * @return Page of a tab as com.spvessel.spacevil.Frame.
     */
    public Frame getTabFrame(Tab tab) {
        return _tabBar.getTabFrame(tab);
    }

    /**
     * Getting content of a page by its tab.
     * 
     * @param tab Tab as com.spvessel.spacevil.Tab.
     * @return Page content as
     *         List&lt;com.spvessel.spacevil.Core.InterfaceBaseItem&gt;.
     */
    public List<InterfaceBaseItem> getTabContent(Tab tab) {
        return _tabBar.getTabContent(tab);
    }

    /**
     * Adding a new tab to the TabView.
     * 
     * @param tab The new tab as com.spvessel.spacevil.Tab.
     */
    public void addTab(Tab tab) {
        if (_tabBar.tabMapView.containsKey(tab))
            return;

        _tabBar.tabMapView.put(tab, tab.view);
        _viewArea.addItem(tab.view);
        _tabBar.addItem(tab);
        tab.eventMousePress.add((sender, args) -> {
            updateLayout();
        });
        tab.eventTabRemove.add(() -> {
            removeTab(tab);
            updateLayout();
        });
    }

    /**
     * Adding new tabs to the TabView.
     * 
     * @param tabs Tab sequence as com.spvessel.spacevil.Tab.
     */
    public void addTabs(Tab... tabs) {
        for (Tab tab : tabs) {
            addTab(tab);
        }
    }

    /**
     * Removing the specified item from TabView.
     * 
     * @param item Item as com.spvessel.spacevil.Core.InterfaceBaseItem.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    @Override
    public boolean removeItem(InterfaceBaseItem item) {
        if (item instanceof Tab) {
            Tab tmpTab = (Tab) item;
            if (_tabBar.tabMapView.containsKey(tmpTab)) {
                return removeTab(tmpTab);
            }
            return false;
        }
        if (item.equals(_tabBar)) {
            _tabBar.removeAllTabs();
        }
        return super.removeItem(item);
    }

    /**
     * Removing a tab by its name.
     * <p>
     * Tab name is ID of UI element. Tab name is uniq but tab text is not.
     * <p>
     * Explanation: the tab text can be a web address, and there can be several open
     * tabs with the same address.
     * 
     * @param tabName Tab name.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    public boolean removeTabByName(String tabName) {
        return _tabBar.removeTabByName(tabName);
    }

    /**
     * Removing a first-found tab by its text.
     * <p>
     * Tab name is ID of UI element. Tab name is uniq but tab text is not.
     * <p>
     * Explanation: the tab text can be a web address, and there can be several open
     * tabs with the same address.
     * 
     * @param tabText Tab text.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    public boolean removeTabByText(String tabText) {
        return _tabBar.removeTabByText(tabText);
    }

    /**
     * Removing a tab.
     * 
     * @param tab Tab as com.spvessel.spacevil.Tab.
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    public boolean removeTab(Tab tab) {
        return _tabBar.removeTab(tab);
    }

    /**
     * Removing all existing tabs.
     * 
     * @return True: if the removal was successful. False: if the removal was
     *         unsuccessful.
     */
    public boolean removeAllTabs() {
        return _tabBar.removeAllTabs();
    }

    /**
     * Adding an item to page by its tab name.
     * <p>
     * Tab name is ID of UI element. Tab name is uniq but tab text is not.
     * <p>
     * Explanation: the tab text can be a web address, and there can be several open
     * tabs with the same address.
     * 
     * @param tabName Tab name.
     * @param item    Item as com.spvessel.spacevil.Core.InterfaceBaseItem.
     */
    public void addItemToTabByName(String tabName, InterfaceBaseItem item) {
        _tabBar.addItemToTabByName(tabName, item);
    }

    /**
     * Adding an item to page by its tab text.
     * <p>
     * Tab name is ID of UI element. Tab name is uniq but tab text is not.
     * <p>
     * Explanation: the tab text can be a web address, and there can be several open
     * tabs with the same address.
     * 
     * @param tabText Tab text.
     * @param item    Item as com.spvessel.spacevil.Core.InterfaceBaseItem.
     */
    public void addItemToTabByText(String tabText, InterfaceBaseItem item) {
        _tabBar.addItemToTabByText(tabText, item);
    }

    /**
     * Adding an item to page by its tab.
     * 
     * @param tab  Tab as com.spvessel.spacevil.Tab.
     * @param item Item as com.spvessel.spacevil.Core.InterfaceBaseItem.
     */
    public void addItemToTab(Tab tab, InterfaceBaseItem item) {
        _tabBar.addItemToTab(tab, item);
    }

    /**
     * Setting style of the TabView.
     * <p>
     * Inner styles: Inner styles: "tabbar", "viewarea".
     * 
     * @param style Style as com.spvessel.spacevil.Decorations.Style.
     */
    @Override
    public void setStyle(Style style) {
        if (style == null)
            return;

        super.setStyle(style);

        Style innerStyle = style.getInnerStyle("tabbar");
        if (innerStyle != null)
            _tabBar.setStyle(innerStyle);
        innerStyle = style.getInnerStyle("viewarea");
        if (innerStyle != null)
            _viewArea.setStyle(innerStyle);
    }
}